---
title: "Data Setup"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dev = 'svg')
options(device = function(file, width, height) {
  svg(tempfile(), width = width, height = height)
})
```

## Introduction

To kick off the Bioinforpedia repo we will need some data to play with. Here we use data from TCGA-LGG, a TCGA project assessing low grade gliomas ([GDC Data Portal][1]).

```{r data.setup, warning=FALSE,message=FALSE,results='hide'}
library(TCGAbiolinks)
library(DESeq2)
library(limma)

# load TCGA LGG data
query <- GDCquery(project = "TCGA-LGG", 
                  data.category = "Gene expression",
                  data.type = "Gene expression quantification",
                  experimental.strategy = "RNA-Seq",
                  platform = "Illumina HiSeq",
                  file.type = "results", 
                  legacy = TRUE)
GDCdownload(query)
lgg <- GDCprepare(query)

# obtain expression and meta data
lgg.exp <- as.data.frame(
  assay(lgg,"raw_count")
)
lgg.pdat <- as.data.frame(
  colData(lgg)
  )


#extract SNP data
query1 <- GDCquery(project = "TCGA-LGG", 
                   data.category = "Simple Nucleotide Variation",
                   data.type = "Masked Somatic Mutation",
                   workflow.type = "MuSE Variant Aggregation and Masking")
GDCdownload(query1)
lgg.snv <- GDCprepare(query1)

#extract Copy Number Data
query2 <- GDCquery(project = "TCGA-LGG", 
                   data.category = "Copy Number Variation",
                   data.type = "Copy Number Segment",
)
GDCdownload(query2)
lgg.cnv <- GDCprepare(query2)

lgg <- list(ExpressionData =lgg.exp,
            PatientData =lgg.pdat,
            ShortVariants =lgg.snv,
            CopyNumberVariants = lgg.cnv)
save(lgg,file = "lgg.rda")
```

## References

1) https://portal.gdc.cancer.gov/projects/TCGA-LGG

[1]: https://portal.gdc.cancer.gov/projects/TCGA-LGG